File: ai_image_viewer/src/AI-Image-Viewer/collab/js/join-work.js
#!/usr/bin/env node

/**
 * AI Image Viewer - Work Joiner Tool (JavaScript)
 * 
 * This tool merges multiple work packages back into a single consolidated package.
 * It combines:
 * - Images from all work packages
 * - AI analysis metadata from all JSON exports
 * - Creates a unified dataset with complete analysis results
 * 
 * Author: AI Image Viewer Team
 */

import { Command } from 'commander';
import fs from 'fs-extra';
import path from 'path';
import { glob } from 'glob';
import chalk from 'chalk';
import tar from 'tar';
import { tmpdir } from 'os';
import { mkdtemp } from 'fs/promises';

class WorkJoiner {
    constructor(packagePaths, outputPath) {
        this.packagePaths = packagePaths.map(p => path.resolve(p));
        this.outputPath = path.resolve(outputPath);
        this.supportedExtensions = ['.jpg', '.jpeg', '.png', '.gif', '.webp', '.bmp'];
        this.htmlFiles = [
            'ai_image.html'
        ];
        // Removed metadataPatterns - now using flexible filename detection
        this.projectRoot = this.findProjectRoot();
        this.tempDir = null; // For extracted .tar.gz files
    }

    findProjectRoot() {
        let current = path.dirname(new URL(import.meta.url).pathname);
        while (current !== path.dirname(current)) {
            const hasAllHtml = this.htmlFiles.every(htmlFile => 
                fs.existsSync(path.join(current, htmlFile))
            );
            if (hasAllHtml) {
                return current;
            }
            current = path.dirname(current);
        }
        throw new Error('Could not find project root with HTML files');
    }

    async detectAndExtractArchives() {
        const extractedPaths = [];
        const archiveFiles = [];
        
        for (const inputPath of this.packagePaths) {
            const stats = await fs.stat(inputPath).catch(() => null);
            
            if (stats && stats.isFile() && inputPath.endsWith('.tar.gz')) {
                // This is a .tar.gz file
                archiveFiles.push(inputPath);
            } else if (stats && stats.isDirectory()) {
                // Check if directory contains .tar.gz files
                const gzFiles = await glob(path.join(inputPath, '*.tar.gz'));
                archiveFiles.push(...gzFiles);
                
                // Also check for existing work-package directories
                const workDirs = await glob(path.join(inputPath, 'work-package-*'));
                for (const workDir of workDirs) {
                    if ((await fs.stat(workDir)).isDirectory()) {
                        extractedPaths.push(workDir);
                    }
                }
            } else if (stats && stats.isDirectory()) {
                // Regular directory
                extractedPaths.push(inputPath);
            }
        }
        
        // Extract .tar.gz files if any found
        if (archiveFiles.length > 0) {
            console.log(chalk.cyan(`🗜️  Found ${archiveFiles.length} compressed packages to extract...`));
            
            // Create temporary directory for extraction
            this.tempDir = await mkdtemp(path.join(tmpdir(), 'ai-image-join-'));
            
            for (const archiveFile of archiveFiles) {
                console.log(chalk.blue(`📦 Extracting ${path.basename(archiveFile)}...`));
                
                try {
                    await tar.x({
                        file: archiveFile,
                        cwd: this.tempDir
                    });
                    console.log(chalk.green(`  ✓ Extracted ${path.basename(archiveFile)}`));
                    
                    // Find extracted work-package directories
                    const extractedDirs = await glob(path.join(this.tempDir, 'work-package-*'));
                    for (const extractedDir of extractedDirs) {
                        if ((await fs.stat(extractedDir)).isDirectory()) {
                            extractedPaths.push(extractedDir);
                        }
                    }
                    
                } catch (error) {
                    console.log(chalk.red(`  ❌ Failed to extract ${path.basename(archiveFile)}: ${error.message}`));
                    continue;
                }
            }
        }
        
        return extractedPaths;
    }

    async cleanupTempFiles() {
        if (this.tempDir && await fs.pathExists(this.tempDir)) {
            await fs.remove(this.tempDir);
            console.log(chalk.green('🧹 Cleaned up temporary files'));
        }
    }

    async validatePackages() {
        const validPackages = [];

        for (const packagePath of this.packagePaths) {
            if (!await fs.pathExists(packagePath)) {
                console.log(chalk.yellow(`⚠️  Warning: Package not found: ${packagePath}`));
                continue;
            }

            const stats = await fs.stat(packagePath);
            if (!stats.isDirectory()) {
                if (path.extname(packagePath) === '.gz' || path.extname(packagePath) === '.zip') {
                    console.log(chalk.blue(`📦 Found compressed package: ${path.basename(packagePath)}`));
                    // TODO: Add extraction logic if needed
                    continue;
                } else {
                    console.log(chalk.yellow(`⚠️  Warning: Expected folder, got file: ${packagePath}`));
                    continue;
                }
            }

            // Check if it's a valid package directory
            const imagesPath = path.join(packagePath, 'images');
            if (!await fs.pathExists(imagesPath)) {
                console.log(chalk.yellow(`⚠️  Warning: No images folder in: ${path.basename(packagePath)}`));
                continue;
            }

            validPackages.push(packagePath);
            console.log(chalk.green(`✓ Valid package: ${path.basename(packagePath)}`));
        }

        return validPackages;
    }

    async collectImages(packages) {
        const allImages = [];
        const seenNames = new Set();

        for (const packagePath of packages) {
            const imagesPath = path.join(packagePath, 'images');
            if (!await fs.pathExists(imagesPath)) {
                continue;
            }

            const patterns = this.supportedExtensions.flatMap(ext => [
                path.join(imagesPath, `*${ext}`),
                path.join(imagesPath, `*${ext.toUpperCase()}`)
            ]);

            const packageImages = [];
            for (const pattern of patterns) {
                const matches = await glob(pattern);
                packageImages.push(...matches);
            }

            console.log(chalk.blue(`📸 Package ${path.basename(packagePath)}: ${packageImages.length} images`));

            for (const imgPath of packageImages) {
                const originalName = path.basename(imgPath);
                if (seenNames.has(originalName)) {
                    // Skip duplicates - first survives
                    console.log(chalk.yellow(`  ⚠️  Skipping duplicate image: ${originalName} (first survives)`));
                    continue;
                } else {
                    allImages.push(imgPath);
                    seenNames.add(originalName);
                }
            }
        }

        return allImages;
    }

    async findMetadataFiles(packages) {
        const metadataFiles = [];

        for (const packagePath of packages) {
            const foundInPackage = [];

            // Find all .json files in the package
            const jsonFiles = await glob(path.join(packagePath, '*.json'));

            // Filter for files containing "metadata" (case-insensitive)
            for (const jsonFile of jsonFiles) {
                const filename = path.basename(jsonFile).toLowerCase();
                if (filename.includes('metadata')) {
                    // Exclude package-manifest.json (not analysis data)
                    if (path.basename(jsonFile) !== 'package-manifest.json') {
                        foundInPackage.push(jsonFile);
                    }
                }
            }

            if (foundInPackage.length > 0) {
                console.log(chalk.blue(`📊 Package ${path.basename(packagePath)}: ${foundInPackage.length} metadata files`));
                for (const metaFile of foundInPackage) {
                    console.log(chalk.white(`  - ${path.basename(metaFile)}`));
                }
                metadataFiles.push(...foundInPackage);
            } else {
                console.log(chalk.yellow(`⚠️  Package ${path.basename(packagePath)}: No metadata files found`));
            }
        }

        return metadataFiles;
    }

    async mergeMetadata(metadataFiles) {
        const merged = {
            version: '2.0-Consolidated',
            timestamp: new Date().toISOString(),
            appName: 'AI Image Viewer - Consolidated Results',
            consolidation_info: {
                source_files: metadataFiles.map(f => path.basename(f)),
                packages_merged: new Set(metadataFiles.map(f => path.basename(path.dirname(f)))).size,
                merge_date: new Date().toISOString(),
                models_used: [],
                total_analyzed_images: 0
            },
            totalImages: 0,
            aiEnabled: true,
            captions: {},
            aiData: {}
        };

        let imageCount = 0;
        const modelsUsed = new Set();

        for (const metaFile of metadataFiles) {
            try {
                const data = await fs.readJSON(metaFile);
                const packageName = path.basename(path.dirname(metaFile));
                
                console.log(chalk.blue(`📋 Processing: ${path.basename(metaFile)}`));

                // Merge captions - first survives for duplicates
                if (data.captions) {
                    for (const [imgName, caption] of Object.entries(data.captions)) {
                        if (!merged.captions[imgName]) {
                            merged.captions[imgName] = caption;
                        }
                        // Skip if duplicate - first survives
                    }
                }

                // Merge AI data - first survives for duplicates
                if (data.aiData) {
                    for (const [imgName, aiData] of Object.entries(data.aiData)) {
                        if (!merged.aiData[imgName]) {
                            merged.aiData[imgName] = aiData;
                        }
                        // Skip if duplicate - first survives

                        // Track models used
                        if (aiData.modelUsed) {
                            modelsUsed.add(aiData.modelUsed);
                        }
                    }
                }

                // Count images
                if (data.totalImages) {
                    imageCount += data.totalImages;
                }

            } catch (error) {
                if (error.name === 'SyntaxError') {
                    console.log(chalk.red(`❌ Error parsing ${path.basename(metaFile)}: ${error.message}`));
                } else {
                    console.log(chalk.yellow(`⚠️  Warning processing ${path.basename(metaFile)}: ${error.message}`));
                }
                continue;
            }
        }

        merged.totalImages = imageCount;
        merged.consolidation_info.models_used = [...modelsUsed];
        merged.consolidation_info.total_analyzed_images = Object.keys(merged.aiData).length;

        return merged;
    }

    async createConsolidatedPackage(images, metadata) {
        console.log(chalk.blue(`📦 Creating consolidated package: ${this.outputPath}`));

        // Create output directory
        await fs.ensureDir(this.outputPath);

        // Copy HTML files from project root
        for (const htmlFile of this.htmlFiles) {
            const srcPath = path.join(this.projectRoot, htmlFile);
            const dstPath = path.join(this.outputPath, htmlFile);
            if (await fs.pathExists(srcPath)) {
                await fs.copy(srcPath, dstPath);
                console.log(chalk.green(`✓ Copied ${htmlFile}`));
            }
        }

        // Copy docs folder if it exists
        const docsSrc = path.join(this.projectRoot, 'docs');
        if (await fs.pathExists(docsSrc)) {
            const docsDst = path.join(this.outputPath, 'docs');
            await fs.copy(docsSrc, docsDst);
            console.log(chalk.green('✓ Copied docs folder'));
        }

        // Create images folder and copy all images
        const imagesDst = path.join(this.outputPath, 'images');
        await fs.ensureDir(imagesDst);

        let copiedCount = 0;
        for (const imgPath of images) {
            const dstPath = path.join(imagesDst, path.basename(imgPath));
            await fs.copy(imgPath, dstPath);
            copiedCount++;
        }

        console.log(chalk.green(`✓ Copied ${copiedCount} images`));

        // Save consolidated metadata
        const metadataPath = path.join(this.outputPath, 'consolidated-metadata.json');
        await fs.writeJSON(metadataPath, metadata, { spaces: 2 });
        console.log(chalk.green('✓ Created consolidated metadata'));

        // Create summary report
        await this.createSummaryReport(metadata);

        return this.outputPath;
    }

    async createSummaryReport(metadata) {
        const reportPath = path.join(this.outputPath, 'consolidation-report.txt');
        
        const reportLines = [
            '🧠 AI Image Viewer - Consolidation Report',
            '='.repeat(50),
            '',
            `Consolidation Date: ${metadata.consolidation_info.merge_date}`,
            `Packages Merged: ${metadata.consolidation_info.packages_merged}`,
            `Source Files: ${metadata.consolidation_info.source_files.length}`,
            `Total Images: ${metadata.totalImages}`,
            `Analyzed Images: ${metadata.consolidation_info.total_analyzed_images}`,
            `Analysis Coverage: ${(metadata.consolidation_info.total_analyzed_images / metadata.totalImages * 100).toFixed(1)}%`,
            ''
        ];

        if (metadata.consolidation_info.models_used.length > 0) {
            reportLines.push('AI Models Used:');
            for (const model of metadata.consolidation_info.models_used) {
                reportLines.push(`  - ${model}`);
            }
            reportLines.push('');
        }

        reportLines.push('Source Metadata Files:');
        for (const sourceFile of metadata.consolidation_info.source_files) {
            reportLines.push(`  - ${sourceFile}`);
        }

        await fs.writeFile(reportPath, reportLines.join('\n'));
        console.log(chalk.green('✓ Created consolidation report'));
    }

    async join() {
        try {
            console.log(chalk.cyan(`🔍 Processing ${this.packagePaths.length} input paths...`));

            // Detect and extract .tar.gz files
            const extractedPaths = await this.detectAndExtractArchives();
            
            // Update package paths with extracted directories
            this.packagePaths = extractedPaths;

            console.log(chalk.cyan(`🔍 Validating ${this.packagePaths.length} packages...`));

            // Validate packages
            const validPackages = await this.validatePackages();
            if (validPackages.length === 0) {
                throw new Error('No valid packages found to merge');
            }

            console.log(chalk.green(`✓ Found ${validPackages.length} valid packages`));

            // Collect all images
            console.log(chalk.cyan('\n📸 Collecting images from packages...'));
            const allImages = await this.collectImages(validPackages);
            console.log(chalk.green(`✓ Collected ${allImages.length} total images`));

            // Find and merge metadata
            console.log(chalk.cyan('\n📊 Processing metadata files...'));
            const metadataFiles = await this.findMetadataFiles(validPackages);

            let mergedMetadata;
            if (metadataFiles.length === 0) {
                console.log(chalk.yellow('⚠️  No metadata files found. Creating package with images only.'));
                mergedMetadata = {
                    version: '2.0-Consolidated',
                    timestamp: new Date().toISOString(),
                    appName: 'AI Image Viewer - Consolidated Results',
                    consolidation_info: {
                        source_files: [],
                        packages_merged: validPackages.length,
                        merge_date: new Date().toISOString(),
                        models_used: [],
                        total_analyzed_images: 0
                    },
                    totalImages: allImages.length,
                    aiEnabled: false,
                    captions: {},
                    aiData: {}
                };
            } else {
                mergedMetadata = await this.mergeMetadata(metadataFiles);
            }

            // Create consolidated package
            console.log(chalk.cyan('\n📦 Creating consolidated package...'));
            const outputPath = await this.createConsolidatedPackage(allImages, mergedMetadata);

            return outputPath;
        } finally {
            // Clean up temporary files
            await this.cleanupTempFiles();
        }
    }
}

// CLI Setup
const program = new Command();

program
    .name('join-work')
    .description('🧠 AI Image Viewer - Work Joiner Tool\n\nMerge multiple work packages into a single consolidated package.')
    .version('1.0.0')
    .argument('<packages...>', 'Work package directories to merge')
    .option('-o, --output <directory>', 'Output folder for consolidated package (default: ./join/consolidated)')
    .option('-f, --force', 'Force overwrite existing output directory')
    .addHelpText('after', `
Examples:
  $ node join-work.js ../split/work-package-*
  $ node join-work.js package1/ package2/ package3/ -o final-results/
  $ npx join-work ../split/* -f
`);

program.parse();

const packages = program.args;
const options = program.opts();

// Validation
if (packages.length < 1) {
    console.error(chalk.red('❌ At least one package must be specified'));
    process.exit(1);
}

// Set default output location
const outputPath = options.output || path.join(process.cwd(), 'join', 'consolidated');

// Main execution
async function main() {
    try {
        // Check if output exists
        if (await fs.pathExists(outputPath) && !options.force) {
            console.log(chalk.yellow(`Output directory ${outputPath} exists.`));
            // In a real CLI, you'd want to prompt for user input
            // For now, we'll just remove it
            await fs.remove(outputPath);
        } else if (await fs.pathExists(outputPath) && options.force) {
            await fs.remove(outputPath);
        }

        console.log(chalk.bold.blue('🧠 AI Image Viewer - Work Joiner Tool'));
        console.log(chalk.blue('='.repeat(43)));

        // Initialize joiner
        const joiner = new WorkJoiner(packages, outputPath);

        // Perform the join
        const resultPath = await joiner.join();

        console.log(chalk.green('\n✅ Successfully created consolidated package!'));
        console.log(chalk.blue(`📁 Output location: ${resultPath}`));

        // Show final summary
        const metadataPath = path.join(resultPath, 'consolidated-metadata.json');
        if (await fs.pathExists(metadataPath)) {
            const metadata = await fs.readJSON(metadataPath);

            console.log(chalk.blue('\n📋 Final Summary:'));
            console.log(chalk.white(`   Total Images: ${metadata.totalImages}`));
            console.log(chalk.white(`   Analyzed Images: ${metadata.consolidation_info?.total_analyzed_images || 0}`));
            console.log(chalk.white(`   Packages Merged: ${packages.length}`));

            const models = metadata.consolidation_info?.models_used || [];
            console.log(chalk.white(`   AI Models Used: ${models.length > 0 ? models.join(', ') : 'None'}`));
        }

        console.log(chalk.green('\n🚀 Next Steps:'));
        console.log(chalk.white('1. Open any HTML file in the consolidated package'));
        console.log(chalk.white('2. Import \'consolidated-metadata.json\' to see all analysis results'));
        console.log(chalk.white('3. Review \'consolidation-report.txt\' for detailed statistics'));

    } catch (error) {
        console.error(chalk.red(`❌ Error: ${error.message}`));
        process.exit(1);
    }
}

main();

File: ai_image_viewer/src/AI-Image-Viewer/collab/python/join_work.py
#!/usr/bin/env python3
"""
AI Image Viewer - Work Joiner Tool

This tool merges multiple work packages back into a single consolidated package.
It combines:
- Images from all work packages
- AI analysis metadata from all JSON exports
- Creates a unified dataset with complete analysis results

Author: AI Image Viewer Team
"""

import os
import shutil
import json
import click
from pathlib import Path
from typing import List, Dict, Any
from datetime import datetime
import glob


class WorkJoiner:
    """Handles merging multiple work packages into a single consolidated package."""
    
    SUPPORTED_EXTENSIONS = {'.jpg', '.jpeg', '.png', '.gif', '.webp', '.bmp'}
    HTML_FILES = [
        'ai_image_viewer.html',
        'ai_image_viewer_efficientnet.html', 
        'ai_image_viewer_mediapipe.html'
    ]
    METADATA_PATTERNS = [
        'ai-image-metadata.json',
        '*-image-metadata.json',
        'efficientnet-image-metadata.json',
        'mediapipe-image-metadata.json',
        '*-analysis.json',
        'metadata.json'
    ]
    
    def __init__(self, package_paths: List[Path], output_path: Path):
        self.package_paths = [Path(p) for p in package_paths]
        self.output_path = Path(output_path)
        self.project_root = self.find_project_root()
        
    def find_project_root(self) -> Path:
        """Find the project root containing HTML files."""
        current = Path(__file__).parent
        while current.parent != current:
            if all((current / html).exists() for html in self.HTML_FILES):
                return current
            current = current.parent
        raise FileNotFoundError("Could not find project root with HTML files")
    
    def validate_packages(self) -> List[Path]:
        """Validate that package paths exist and contain expected content."""
        valid_packages = []
        
        for package_path in self.package_paths:
            if not package_path.exists():
                click.echo(f"⚠️  Warning: Package not found: {package_path}")
                continue
                
            if package_path.is_file():
                # Handle compressed packages
                if package_path.suffix in {'.zip', '.tar', '.gz'}:
                    click.echo(f"📦 Found compressed package: {package_path.name}")
                    # TODO: Add extraction logic if needed
                    continue
                else:
                    click.echo(f"⚠️  Warning: Expected folder, got file: {package_path}")
                    continue
            
            # Check if it's a valid package directory
            images_path = package_path / "images"
            if not images_path.exists():
                click.echo(f"⚠️  Warning: No images folder in: {package_path}")
                continue
            
            valid_packages.append(package_path)
            click.echo(f"✓ Valid package: {package_path.name}")
        
        return valid_packages
    
    def collect_images(self, packages: List[Path]) -> List[Path]:
        """Collect all image files from packages."""
        all_images = []
        seen_names = set()
        
        for package in packages:
            images_path = package / "images"
            if not images_path.exists():
                continue
                
            package_images = []
            for ext in self.SUPPORTED_EXTENSIONS:
                package_images.extend(images_path.glob(f"*{ext}"))
                package_images.extend(images_path.glob(f"*{ext.upper()}"))
            
            click.echo(f"📸 Package {package.name}: {len(package_images)} images")
            
            for img in package_images:
                # Handle duplicate names by adding package prefix
                original_name = img.name
                if original_name in seen_names:
                    new_name = f"{package.name}_{original_name}"
                    click.echo(f"  ⚠️  Renaming duplicate: {original_name} → {new_name}")
                    # Create a temporary renamed copy
                    temp_path = img.parent / new_name
                    shutil.copy2(img, temp_path)
                    all_images.append(temp_path)
                    seen_names.add(new_name)
                else:
                    all_images.append(img)
                    seen_names.add(original_name)
        
        return all_images
    
    def find_metadata_files(self, packages: List[Path]) -> List[Path]:
        """Find all metadata JSON files in packages."""
        metadata_files = []
        
        for package in packages:
            found_in_package = []
            
            # Search for metadata files using patterns
            for pattern in self.METADATA_PATTERNS:
                matches = list(package.glob(pattern))
                found_in_package.extend(matches)
            
            # Remove duplicates
            found_in_package = list(set(found_in_package))
            
            if found_in_package:
                click.echo(f"📊 Package {package.name}: {len(found_in_package)} metadata files")
                for meta_file in found_in_package:
                    click.echo(f"  - {meta_file.name}")
                metadata_files.extend(found_in_package)
            else:
                click.echo(f"⚠️  Package {package.name}: No metadata files found")
        
        return metadata_files
    
    def merge_metadata(self, metadata_files: List[Path]) -> Dict[str, Any]:
        """Merge multiple metadata JSON files into a single consolidated metadata."""
        merged = {
            "version": "2.0-Consolidated",
            "timestamp": datetime.now().isoformat(),
            "appName": "AI Image Viewer - Consolidated Results",
            "consolidation_info": {
                "source_files": [f.name for f in metadata_files],
                "packages_merged": len(set(f.parent.name for f in metadata_files)),
                "merge_date": datetime.now().isoformat()
            },
            "totalImages": 0,
            "aiEnabled": True,
            "captions": {},
            "aiData": {}
        }
        
        image_count = 0
        models_used = set()
        
        for meta_file in metadata_files:
            try:
                with open(meta_file, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                
                package_name = meta_file.parent.name
                click.echo(f"📋 Processing: {meta_file.name}")
                
                # Merge captions
                if 'captions' in data:
                    for img_name, caption in data['captions'].items():
                        # Handle duplicate image names
                        key = img_name
                        if key in merged['captions']:
                            key = f"{package_name}_{img_name}"
                        merged['captions'][key] = caption
                
                # Merge AI data
                if 'aiData' in data:
                    for img_name, ai_data in data['aiData'].items():
                        # Handle duplicate image names
                        key = img_name
                        if key in merged['aiData']:
                            key = f"{package_name}_{img_name}"
                        merged['aiData'][key] = ai_data
                        
                        # Track models used
                        if 'modelUsed' in ai_data:
                            models_used.add(ai_data['modelUsed'])
                
                # Count images
                if 'totalImages' in data:
                    image_count += data['totalImages']
                
            except json.JSONDecodeError as e:
                click.echo(f"❌ Error parsing {meta_file.name}: {e}")
                continue
            except Exception as e:
                click.echo(f"⚠️  Warning processing {meta_file.name}: {e}")
                continue
        
        merged['totalImages'] = image_count
        merged['consolidation_info']['models_used'] = list(models_used)
        merged['consolidation_info']['total_analyzed_images'] = len(merged['aiData'])
        
        return merged
    
    def create_consolidated_package(self, images: List[Path], metadata: Dict[str, Any]) -> Path:
        """Create the final consolidated package."""
        click.echo(f"📦 Creating consolidated package: {self.output_path}")
        
        # Create output directory
        self.output_path.mkdir(parents=True, exist_ok=True)
        
        # Copy HTML files from project root
        for html_file in self.HTML_FILES:
            src_path = self.project_root / html_file
            dst_path = self.output_path / html_file
            if src_path.exists():
                shutil.copy2(src_path, dst_path)
                click.echo(f"✓ Copied {html_file}")
        
        # Copy docs folder if it exists
        docs_src = self.project_root / "docs"
        if docs_src.exists():
            docs_dst = self.output_path / "docs"
            shutil.copytree(docs_src, docs_dst, dirs_exist_ok=True)
            click.echo(f"✓ Copied docs folder")
        
        # Create images folder and copy all images
        images_dst = self.output_path / "images"
        images_dst.mkdir(exist_ok=True)
        
        copied_count = 0
        for img_path in images:
            dst_path = images_dst / img_path.name
            shutil.copy2(img_path, dst_path)
            copied_count += 1
        
        click.echo(f"✓ Copied {copied_count} images")
        
        # Save consolidated metadata
        metadata_path = self.output_path / "consolidated-metadata.json"
        with open(metadata_path, 'w', encoding='utf-8') as f:
            json.dump(metadata, f, indent=2, ensure_ascii=False)
        
        click.echo(f"✓ Created consolidated metadata")
        
        # Create summary report
        self.create_summary_report(metadata)
        
        return self.output_path
    
    def create_summary_report(self, metadata: Dict[str, Any]):
        """Create a human-readable summary report."""
        report_path = self.output_path / "consolidation-report.txt"
        
        with open(report_path, 'w', encoding='utf-8') as f:
            f.write("🧠 AI Image Viewer - Consolidation Report\n")
            f.write("=" * 50 + "\n\n")
            
            f.write(f"Consolidation Date: {metadata['consolidation_info']['merge_date']}\n")
            f.write(f"Packages Merged: {metadata['consolidation_info']['packages_merged']}\n")
            f.write(f"Source Files: {len(metadata['consolidation_info']['source_files'])}\n")
            f.write(f"Total Images: {metadata['totalImages']}\n")
            f.write(f"Analyzed Images: {metadata['consolidation_info']['total_analyzed_images']}\n")
            f.write(f"Analysis Coverage: {metadata['consolidation_info']['total_analyzed_images']/metadata['totalImages']*100:.1f}%\n\n")
            
            if metadata['consolidation_info']['models_used']:
                f.write("AI Models Used:\n")
                for model in metadata['consolidation_info']['models_used']:
                    f.write(f"  - {model}\n")
                f.write("\n")
            
            f.write("Source Metadata Files:\n")
            for source_file in metadata['consolidation_info']['source_files']:
                f.write(f"  - {source_file}\n")
        
        click.echo(f"✓ Created consolidation report")
    
    def join(self) -> Path:
        """Main method to join work packages."""
        click.echo(f"🔍 Validating {len(self.package_paths)} packages...")
        
        # Validate packages
        valid_packages = self.validate_packages()
        if not valid_packages:
            raise click.ClickException("No valid packages found to merge")
        
        click.echo(f"✓ Found {len(valid_packages)} valid packages")
        
        # Collect all images
        click.echo(f"\n📸 Collecting images from packages...")
        all_images = self.collect_images(valid_packages)
        click.echo(f"✓ Collected {len(all_images)} total images")
        
        # Find and merge metadata
        click.echo(f"\n📊 Processing metadata files...")
        metadata_files = self.find_metadata_files(valid_packages)
        
        if not metadata_files:
            click.echo("⚠️  No metadata files found. Creating package with images only.")
            merged_metadata = {
                "version": "2.0-Consolidated",
                "timestamp": datetime.now().isoformat(),
                "appName": "AI Image Viewer - Consolidated Results",
                "consolidation_info": {
                    "source_files": [],
                    "packages_merged": len(valid_packages),
                    "merge_date": datetime.now().isoformat(),
                    "models_used": [],
                    "total_analyzed_images": 0
                },
                "totalImages": len(all_images),
                "aiEnabled": False,
                "captions": {},
                "aiData": {}
            }
        else:
            merged_metadata = self.merge_metadata(metadata_files)
        
        # Create consolidated package
        click.echo(f"\n📦 Creating consolidated package...")
        output_path = self.create_consolidated_package(all_images, merged_metadata)
        
        return output_path


@click.command()
@click.argument('packages', nargs=-1, required=True,
                type=click.Path(exists=True, path_type=Path))
@click.option('--output', '-o',
              type=click.Path(path_type=Path),
              default=None,
              help='Output folder for consolidated package (default: ./join/consolidated)')
@click.option('--force', '-f',
              is_flag=True,
              help='Force overwrite existing output directory')
@click.version_option(version='1.0.0', prog_name='AI Image Viewer Work Joiner')
def main(packages: List[Path], output: Path, force: bool):
    """
    🧠 AI Image Viewer - Work Joiner Tool
    
    Merge multiple work packages into a single consolidated package.
    Combines images and AI analysis metadata from team collaboration.
    
    PACKAGES: One or more work package directories or compressed files to merge
    
    Examples:
    
    \b
    # Join all packages in split folder
    python join_work.py split/work-package-*
    
    \b
    # Join specific packages with custom output
    python join_work.py package1/ package2/ package3/ -o final-results/
    
    \b
    # Force overwrite existing output
    python join_work.py split/* -f
    """
    
    if len(packages) < 1:
        raise click.BadParameter("At least one package must be specified")
    
    # Set default output location
    if output is None:
        output = Path(__file__).parent / "join" / "consolidated"
    
    # Check if output exists
    if output.exists() and not force:
        if click.confirm(f"Output directory {output} exists. Overwrite?"):
            shutil.rmtree(output)
        else:
            click.echo("❌ Aborted by user")
            return
    elif output.exists() and force:
        shutil.rmtree(output)
    
    try:
        click.echo("🧠 AI Image Viewer - Work Joiner Tool")
        click.echo("=" * 43)
        
        # Initialize joiner
        joiner = WorkJoiner(packages, output)
        
        # Perform the join
        result_path = joiner.join()
        
        click.echo(f"\n✅ Successfully created consolidated package!")
        click.echo(f"📁 Output location: {result_path}")
        
        # Show final summary
        metadata_path = result_path / "consolidated-metadata.json"
        if metadata_path.exists():
            with open(metadata_path, 'r') as f:
                metadata = json.load(f)
            
            click.echo(f"\n📋 Final Summary:")
            click.echo(f"   Total Images: {metadata['totalImages']}")
            click.echo(f"   Analyzed Images: {metadata.get('consolidation_info', {}).get('total_analyzed_images', 0)}")
            click.echo(f"   Packages Merged: {len(packages)}")
            
            if 'models_used' in metadata.get('consolidation_info', {}):
                models = metadata['consolidation_info']['models_used']
                click.echo(f"   AI Models Used: {', '.join(models) if models else 'None'}")
        
        click.echo(f"\n🚀 Next Steps:")
        click.echo(f"1. Open any HTML file in the consolidated package")
        click.echo(f"2. Import 'consolidated-metadata.json' to see all analysis results")
        click.echo(f"3. Review 'consolidation-report.txt' for detailed statistics")
        
    except Exception as e:
        click.echo(f"❌ Error: {e}", err=True)
        raise click.Abort()


if __name__ == '__main__':
    main()